### PR TITLE
add localised style hosting for qwant IA

### PR DESCRIPTION
**Pull request by Julien-laville**
add pre-built style for qwant services available at /style.json
lang param can be passed as ?lang={2 letters lang code}

### CODE REVIEW

File: bin/middlewares/map_style.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  let builtStyle = styleBuilder(qwantStyle, options)
+
+  return function (req, res) {
+    /* get lang routine */
+    let langParameter = req.query.lang
+    let matchedLanguage = languages.defaultLanguage
+    if(langParameter) {
+      matchedLanguage = languages.supportedLanguages.find((supportedLanguage) => supportedLanguage.code === langParameter)
+      if(!matchedLanguage) {
+        res.send(400)
```
Comment (amatissart) [inline bin/middlewares/map_style.js]: `res.send` with a status code is deprecated.  
And it is actually an error to call `res.setHeader` after `res.send` so we should return here.  
Maybe we could include an explicit json error in the response too.

File: bin/middlewares/map_style.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  /* pre-build style on server start */
+  let options = {
+    output: 'production', // 'debug' | 'production' | 'omt'
+    outPath : __dirname + '/../public/mapstyle',
+    i18n : true,
+    icons : true,
+    pixelRatios : [1,2],
+    styleDir : path.resolve(path.join(__dirname, '..', '..', 'node_modules', '@qwant' , 'qwant-basic-gl-style')),
+    conf : {tileserver_base : '{tileserver_base}', tileserver_poi : '{tileserver_poi}', spriteserver : '{spriteserver}', fontserver : '{fontserver}'},
+    ignoreSprite : true
```
Comment (amatissart) [inline bin/middlewares/map_style.js]: Because of this flag, `sprite` property will be removed from the style object.  
So the spriteserver is not present in the result :(

File: local_modules/mapbox_style_configure/index.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+module.exports = function configure(style, mapStyleConfig, lang) {
+  return JSON.parse(style
+    .replace(/\{locale\}/g, lang)
+    .replace('"{tileserver_base}"', mapStyleConfig.baseMapUrl)
+    .replace('"{tileserver_poi}"', mapStyleConfig.poiMapUrl)
+    .replace('{spriteserver}', mapStyleConfig.spritesUrl)
+    .replace('{fontserver}', mapStyleConfig.fontsUrl)
```
Comment (amatissart) [inline local_modules/mapbox_style_configure/index.js]: `spriteserver` and `fontserver` need to be absolute urls.  
We could probably build it using `req` properties (similar to  "src/adapters/scene_config.js" )

File: bin/app.js [hunk 4]
```diff
@@ -59,6 +60,8 @@ function App(config) {
     maxAge: config.statics.maxAge
   }))
 
+  app.use('/style.json', new mapStyle(config, languages))
```
Comment (amatissart) [inline bin/app.js]: I think we should use express compression on this endpoint.  
The gzipped style is 8KB, compared to 100KB without compression

File: bin/middlewares/map_style.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  let builtStyle = styleBuilder(qwantStyle, options)
+
+  return function (req, res) {
+    /* get lang routine */
+    let langParameter = req.query.lang
```
Comment (amatissart) [inline bin/middlewares/map_style.js]: At this point, it could be a good idea to add documentation about this kind of endpoints and parameters. In the readme for example.  
Especially as this one will be used by other services.

File: bin/middlewares/map_style.js (hunk 1 above)
Comment (Julien-laville) [reply bin/middlewares/map_style.js]: i'll use res.status
update to a json like {error : Unsupported language code : #lang code}
and return to avoid another send :)

File: bin/middlewares/map_style.js (hunk 2 above)
Comment (Julien-laville) [reply bin/middlewares/map_style.js]: i think,i'll keep the old boolean (sprite exist) for transform call & use build sprite for the spritezero call.

File: local_modules/mapbox_style_configure/index.js (hunk 3 above)
Comment (Julien-laville) [reply local_modules/mapbox_style_configure/index.js]: i think 
{ spritesUrl : `${req.get('host')}${config.mapStyle.spritesUrl}` ,
 fontsUrl : `${req.get('host')}${config.mapStyle.fontsUrl}`
}
is enough

File: bin/app.js (hunk 4 above)
Comment (Julien-laville) [reply bin/app.js]: sure, i'll look for the how to

File: bin/middlewares/map_style.js (hunk 5 above)
Comment (Julien-laville) [reply bin/middlewares/map_style.js]: compression (the most used middleare) enable gzip for all request, it take a filter param but i think it's dirty way. 
or i can enable compression for all request

File: bin/app.js (hunk 4 above)
Comment (Julien-laville) [reply bin/app.js]: compression (the most used middleare) enable gzip for all request, it take a filter param but i think it's dirty way. 
or i can enable compression for all request

File: bin/middlewares/map_style.js (hunk 5 above)
Comment (Julien-laville) [reply bin/middlewares/map_style.js]: done

File: bin/middlewares/map_style.js (hunk 2 above)
Comment (amatissart) [reply bin/middlewares/map_style.js]: cool ! please create a PR on map-style-builder

File: bin/middlewares/map_style.js (hunk 5 above)
Comment (amatissart) [reply bin/middlewares/map_style.js]: could we describe the default behavior (without `lang` too) ?

File: bin/middlewares/map_style.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        res.send({error : `Unsupported language code : ${langParameter}`})
+        return
+      }
+    } else {
+      matchedLanguage = languages.defaultLanguage
+    }
+
+    /* config rewriting */
+    let host = req.get('host')
+    let protocol = req.protocol
```
Comment (amatissart) [inline bin/middlewares/map_style.js]: `req.protocol` will use the `X-Forwarded-Proto` header to determine protocol.  
But it seems that in our docker swarm environments these header will be overridden with `http` by internal services. 
So it is not reliable, and we should probably always use `https` here.

File: bin/app.js (hunk 4 above)
Comment (Julien-laville) [reply bin/app.js]: didn't know about it 
the force is fronce in this one

File: bin/middlewares/map_style.js (hunk 2 above)
Comment (Julien-laville) [reply bin/middlewares/map_style.js]: https://github.com/QwantResearch/map-style-builder/pull/23

File: bin/middlewares/map_style.js (hunk 5 above)
Comment (Julien-laville) [reply bin/middlewares/map_style.js]: yep added default param to doc

File: bin/middlewares/map_style.js (hunk 6 above)
Comment (Julien-laville) [reply bin/middlewares/map_style.js]: hmmm ok. i'm impressed

Comment (Julien-laville) [reply bin/middlewares/map_style.js]: it will be harder to test on local by replacing style object by url

Comment (amatissart) [reply bin/middlewares/map_style.js]: we could work around that, and use `http` when `host === 'localhost'` ?
Or define a new setting with default `https` that we could override in local environments ?

Comment (Julien-laville) [reply bin/middlewares/map_style.js]: :S i prefer not and say it's exclusive for external securised services

File: bin/middlewares/map_style.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+      }
+    } else {
+      matchedLanguage = languages.defaultLanguage
+    }
+
+    /* config rewriting */
+    let host = req.get('host')
+    let urls = {
+      spritesUrl : `https://${host}${config.mapStyle.spritesUrl}` ,
+      fontsUrl : `https://${host}${config.mapStyle.fontsUrl}`
```
Comment (amatissart) [inline bin/middlewares/map_style.js]: Arf, I just realized that we need to handle `baseUrl` here too :(

I think we should try to refactor `toAbsoluteUrl` from "src/adapters/scene_config.js" to be use both on client and server. But that will be for another PR.
